src/librbd/api/Mirror.cc

```diff
@@ -2098,6 +2345,1689 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+    on_finish = new C_SaferCond;
+    image_promote(image_ctxs[i], group_ioctx.get_id(), group_id, group_snap_id,
+                  force, &snap_ids[i], on_finish);
+    on_finishes[i] = on_finish;
+  }
+
+  for (size_t i = 0; i < image_ctxs.size(); i++) {
+    r = 0;
+    if (on_finishes[i]) {
+      r = on_finishes[i]->wait();
+      delete on_finishes[i];
+    }
+    if (r < 0) {
+      if (ret_code == 0) {
+        ret_code = r;
+      }
+    } else {
+      group_snap.snaps[i].snap_id = snap_ids[i];
+    }
+  }
+
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+  if (ret_code < 0) {
+    // undo
+    ldout(cct, 20) << "undoing group promote: " << ret_code << dendl;
+    remove_interim_snapshots(group_ioctx, group_header_oid, &image_ctxs, &group_snap);
+    std::fill(snap_ids.begin(), snap_ids.end(), CEPH_NOSNAP);
+    group_snap.snaps.clear();
+    if (!quiesce_requests.empty()) {
+      util::notify_unquiesce(image_ctxs, quiesce_requests);
+    }
+    close_images(&image_ctxs);
+
+    r = prepare_group_images(group_ioctx, group_id, &image_ctxs,
+                             &group_snap, quiesce_requests,
+                             cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED,
+                             flags);
+    if (r != 0) {
+      return r;
+    }
+    for (size_t i = 0; i < image_ctxs.size(); ++i) {
+      ImageCtx *ictx = image_ctxs[i];
+      ldout(cct, 20) << "demoting individual image: "
+                     << ictx->name.c_str() << dendl;
+      on_finish = new C_SaferCond;
+      image_demote(ictx, group_ioctx.get_id(), group_id,
+                   group_snap_id, &snap_ids[i], on_finish);
+      on_finishes[i] = on_finish;
+    }
+
+    for (int i = 0, n = image_ctxs.size(); i < n; ++i) {
+      if (!on_finishes[i]) {
+        continue;
+      }
+      r = on_finishes[i]->wait();
+      delete on_finishes[i];
+      if (r < 0) {
+        lderr(cct) << "failed demoting image: " << cpp_strerror(r) << dendl;
+        // just report error, but don't abort the process
+      } else{
+        group_snap.snaps[i].snap_id = snap_ids[i];
+      }
+    }
+  }
+
+  group_snap.state = cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE;
+  r = cls_client::group_snap_set(&group_ioctx, group_header_oid, group_snap);
+  if (r < 0) {
+    lderr(cct) << "failed to update group snapshot metadata: "
+               << cpp_strerror(r) << dendl;
+  }
+
+  if (!quiesce_requests.empty()) {
+    util::notify_unquiesce(image_ctxs, quiesce_requests);
+  }
+
+  close_images(&image_ctxs);
+
+  return ret_code;
+}
+
+template <typename I>
+int Mirror<I>::group_demote(IoCtx& group_ioctx,
+                            const char *group_name,
+                            uint32_t flags) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  ldout(cct, 20) << "io_ctx=" << &group_ioctx
+		 << " group name " << group_name << dendl;
+
+  std::string group_id;
+  int r = cls_client::dir_get_id(&group_ioctx, RBD_GROUP_DIRECTORY,
```

Comment:
Lines 3051 to 3075 can be moved into a separate class (say GetGroupInfoRequest)  similar to mirror::GetInfoRequest and reused in multiple functions.